File: ch1/caesar_enc.py
# 16. 파이썬으로 시저 암호 만들기
def encrypt(txt, shift):
    enc_txt = ""
    code_a = ord("A")
    code_z = ord("Z")
    for idx in range(0, len(txt)):
        txt_ord = ord(txt[idx])
        if code_a < txt_ord < code_z:
            txt_ord = txt_ord+shift
        enc_txt += chr(txt_ord)
    return enc_txt

# 17. 시저 함수 조금 더 우아하게 만들기
def encrypt2(txt, shift):
    conv = lambda n: chr(ord(n)+shift)
    enc1 = lambda n: conv(n) if 'A' <= n <= 'Z' else n
    return ''.join([enc1(n) for n in txt])
originTxt = "I LOVE RUST"
shift = 3
encTxt = encrypt2(originTxt, shift)
decTxt = encrypt2(encTxt, -shift)
print("origin = {}".format(originTxt))
print("encrypt = {}".format(encTxt))
print("decrypt = {}".format(decTxt))

File: ch1/caesar_enc.rs
// 16. 러스트로 시저 암호 만들기
fn encrypt(txt: &str, shift: i16) -> String{
    // 'A'와 'Z'의 문자코드를 i16 타입으로 획득
    let code_a = 'A' as i16;
    let code_z = 'Z' as i16;
    // 결과를 대입할 변수를 선언
    let mut enc_txt = String::new();
    for idx in txt.chars() {
        let mut txt_code = idx as i16;
        if code_a < txt_code && txt_code < code_z {
            txt_code += shift;
        }
        enc_txt.push((txt_code as u8) as char);
    }
    return enc_txt;
}

// 19. 시저 함수 조금 더 우아하게 만들기
fn encrypt2(txt: &str, shift: i16) -> String {
    let is_az = |n| 'Z' > n && n > 'A';
    let conv = |n| ((n+shift) as u8) as char;
    let enc1 = |n| if is_az(n) {conv(n as i16)} else {n};

    txt.chars().map(|n| enc1(n)).collect()
}
fn main(){
    // 16
    println!("16");
    let txt = "I LOVE RUST";
    let shift = 3;
    let enc = encrypt(txt, shift);
    let dec = encrypt(&enc, -shift);
    println!("{} => {} => {}", txt, enc, dec);

    // 19
    println!("19");
    let txt = "I LOVE RUST";
    let shift = 3;
    let enc = encrypt2(txt, shift);
    let dec = encrypt2(&enc, -shift);
    println!("{} => {} => {}", txt, enc, dec);
}